internal/provisioner/common/bundle.go

```diff
@@ -0,0 +1,160 @@
+			return ctrl.Result{}, UpdateStatusUnpackFailing(u, fmt.Errorf("get content URL: %v", err))
+		}
+
+		UpdateStatusUnpacked(u, unpackResult, contentURL)
+		return ctrl.Result{}, nil
+	default:
+		return ctrl.Result{}, UpdateStatusUnpackFailing(u, fmt.Errorf("unknown unpack state %q", unpackResult.State))
+	}
+}
+
+func (r *ProvisionerBundleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+	l.V(1).Info("starting reconciliation")
+	defer l.V(1).Info("ending reconciliation")
+	bundle := &rukpakv1alpha1.Bundle{}
+	if err := r.Get(ctx, req.NamespacedName, bundle); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	u := updater.NewBundleUpdater(r.Client)
+	defer func() {
+		if err := u.Apply(ctx, bundle); err != nil {
+			l.Error(err, "failed to update status")
+		}
+	}()
+	u.UpdateStatus(updater.EnsureObservedGeneration(bundle.Generation))
+
+	finalizerResult, err := r.Finalizers.Finalize(ctx, bundle)
+	if err != nil {
+		u.UpdateStatus(
+			updater.EnsureResolvedSource(nil),
+			updater.EnsureContentURL(""),
+			updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+			updater.EnsureCondition(metav1.Condition{
+				Type:    rukpakv1alpha1.TypeUnpacked,
+				Status:  metav1.ConditionUnknown,
+				Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+				Message: err.Error(),
+			}),
+		)
+		return ctrl.Result{}, err
+	}
+	var (
+		finalizerUpdateErrs []error
+	)
+	// Update the status subresource before updating the main object. This is
+	// necessary because, in many cases, the main object update will remove the
+	// finalizer, which will cause the core Kubernetes deletion logic to
+	// complete. Therefore, we need to make the status update prior to the main
+	// object update to ensure that the status update can be processed before
+	// a potential deletion.
+	if finalizerResult.StatusUpdated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Status().Update(ctx, bundle))
+	}
+	if finalizerResult.Updated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Update(ctx, bundle))
+	}
+	if finalizerResult.Updated || finalizerResult.StatusUpdated || !bundle.GetDeletionTimestamp().IsZero() {
+		err := apimacherrors.NewAggregate(finalizerUpdateErrs)
+		if err != nil {
+			u.UpdateStatus(
+				updater.EnsureResolvedSource(nil),
+				updater.EnsureContentURL(""),
+				updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+				updater.EnsureCondition(metav1.Condition{
+					Type:    rukpakv1alpha1.TypeUnpacked,
+					Status:  metav1.ConditionUnknown,
+					Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+					Message: err.Error(),
+				}),
+			)
+		}
+		return ctrl.Result{}, err
+	}
+
+	unpackResult, err := r.Unpacker.Unpack(ctx, bundle)
+	if err != nil {
+		return ctrl.Result{}, UpdateStatusUnpackFailing(&u, fmt.Errorf("source bundle content: %v", err))
+	}
+
+	if r.ParseUnpackState == nil {
+		r.ParseUnpackState = defaultParseUnpackStateFunc
+	}
```

Comment:
It would probably be better not to mutate the state of `r` during reconcile (this function might be called concurrently). Perhaps we could have a function that initializes `ProvisionerBundleReconciler` (e.g. `NewProvisionerBundleReconciler`) where we could do this check/set the default there, prior to the reconciler actually being used?

Reply:
Leaving this as unresolved so you can see how you feel about this most recent implementation coming through now.